pkg/agent/flowexporter/exporter/exporter.go

```diff
@@ -301,15 +301,10 @@ func (exp *flowExporter) sendFlowRecords() error {
 			exp.numDataSetsSent = exp.numDataSetsSent + 1
 
 			if flowexporter.IsConnectionDying(&record.Conn) {
-				// If the connection is in dying state or connection is not in conntrack table,
-				// we will delete the flow records from records map.
-				klog.V(2).Infof("Deleting the inactive flow records with key: %v from record map", key)
-				if err := exp.flowRecords.DeleteFlowRecordWithoutLock(key); err != nil {
-					return err
-				}
-				if err := exp.conntrackConnStore.SetExportDone(key); err != nil {
-					return err
-				}
+				// If the connection is in dying state or connection is not in conntrack
+				// table, we set the DyingAndDoneExport flag to do the deletion later.
+				record.DyingAndDoneExport = true
```

Comment:
as a future improvement, maybe we should just change `ForAllFlowRecordsDo` so that `updateOrSendFlowRecord` uses a flow record pointer instead of a copy of stored flow record. This whole code is executed with the lock any way.

Reply:
Agreed. Or I was thinking to store the pointer to the record in the record map instead of value.
As this code is being refactored and the flow record map would be mostly removed in PR #2360, I tried to keep the changes to a minimum.